distbench_test_sequencer_test.cc

```diff
@@ -159,6 +159,93 @@ TEST(DistBenchTestSequencer, NonEmptyGroup) {
   ASSERT_EQ(s2_1_echo->second.successful_rpc_samples_size(), 10);
 }
 
+TEST(DistBenchTestSequencer, Overload) {
+  DistBenchTester tester;
+  ASSERT_OK(tester.Initialize(2));
+  const std::string proto = R"(
+tests {
+  max_thread_backlog: 4
+  protocol_driver_options {
+    name: 'default_protocol_driver_options'
+    protocol_name: 'grpc'
+    server_settings {
+      name: 'server_type'
+      string_value: 'handoff'
+    }
+  }
+  services {
+    name: "client"
+    count: 1
+  }
+  services {
+    name: "server"
+    count: 1
+  }
+  action_lists {
+    name: "client"
+    action_names: "run_overloading_queries"
+  }
+  actions {
+    name: "run_overloading_queries"
+    rpc_name: "overload_query"
+    iterations {
+      max_iteration_count: 1000
+      max_parallel_iterations: 100
+    }
+  }
+  rpc_descriptions {
+    name: "overload_query"
+    client: "client"
+    server: "server"
+  }
+  action_lists {
+    name: "overload_query"
+    action_names: "simulate_overload"
+  }
+  actions {
+    name: "simulate_overload"
+    activity_config_name: "simulate_overload_activity"
+  }
+  activity_configs {
+    name: "simulate_overload_activity"
+    activity_settings {
+      name: "activity_func"
+      string_value: "SleepFor"
+    }
+    activity_settings {
+      name: "duration_us"
+      int64_value: 10000000
+    }
+  }
+})";
+  auto test_sequence = ParseTestSequenceTextProto(proto);
+  ASSERT_TRUE(test_sequence.ok());
+
+  TestSequenceResults results;
+  auto context = CreateContextWithDeadline(/*max_time_s=*/75);
+  grpc::Status status = tester.test_sequencer_stub->RunTestSequence(
+      context.get(), *test_sequence, &results);
+  ASSERT_OK(status);
+
+  auto& test_results = results.test_results(0);
+  auto service_logs = test_results.service_logs();
+  ASSERT_EQ(service_logs.instance_logs().size(), 1);
+  EXPECT_EQ(service_logs.instance_logs().begin()->first, "client/0");
+  auto& log = service_logs.instance_logs().begin()->second;
+  ASSERT_EQ(log.engine_error_message(),
+            "RESOURCE_EXHAUSTED: Too many threads running");
+  ASSERT_EQ(log.peer_logs_size(), 1);
+  auto peer_log = log.peer_logs().begin()->second.rpc_logs().at(0);
+  EXPECT_LE(peer_log.successful_rpc_samples_size(), 10);
```

Comment:
Still flacky even on my desktop, maybe increase to 30 and something like that for clarity
```cpp
  size_t min_nb_success = 4;
  size_t max_nb_success = 30;
  EXPECT_LE(peer_log.successful_rpc_samples_size(), max_nb_success);
  EXPECT_GE(peer_log.failed_rpc_samples_size(), 100 - max_nb_success);
  EXPECT_LE(peer_log.failed_rpc_samples_size(), 100 - min_nb_success);
```

Reply:
I've rewritten the code to use a WaitWithTimeout, which helps the unit tests without potentially deadlocking in the general case.